### PR TITLE
ON-HOLD: Upgrade operator-sdk v0.17.1

### LINKED ISSUE #167: Upgrade to latest version on operator-sdk
**Describe the solution you'd like**
Upgrade the code to support the latest operator-sdk.

**Describe alternatives you've considered**
Stick with the old and create more technical debt.

**What version of eunomia are you using?**
eunomia version: 0.0.x

### PR DESCRIPTION
**Pull request by Smiley73**
**Description**

Upgrade Operator SDK to v0.17.1

Fixes #167 

**Type of change**

* New feature (non-breaking change which adds functionality)

**Checklist**

- [X] Unit tests and e2e tests updated
- [X] Documentation updated

### CODE REVIEW

**Comment by seanmalloy** [comment]
@Smiley73 please fix the conflict with `.travis.yml` when you have some time. Thanks!

**Comment by Smiley73** [comment]
@seanmalloy yeah I just saw it. Will take care of it.

**Comment by Smiley73** [comment]
@seanmalloy / @MeganMoran this PR is ready for review now. It passes on OpenShift 3.11 and K8s 1.18.0

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -77,9 +77,14 @@ Here are vendor instructions on how to install the necessary components on Linux
 - [make](http://ftp.gnu.org/gnu/make/) (a tarball to download)
 - [shfmt](https://github.com/mvdan/sh#shfmt)
 
-#### Installing operator-sdk v0.8.1 from GitHub release
+#### Installing operator-sdk v0.17.1 from GitHub release
 
-Unfortunately, operator-sdk version 0.8.1 isn't available via package managers, so you have to install it from the GitHub release as described in the [documentation](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/doc/user/install-operator-sdk.md).
+Unfortunately, operator-sdk version 0.17.1 isn't available via package managers, so you have to install it from the GitHub release as described in the [documentation](https://github.com/operator-framework/operator-sdk/blob/v0.17.1/doc/user/install-operator-sdk.md).
+
+The script `scripts/install-operator-sdk.sh ` will do this for you on Linux and MacOs:
```
Comment (MeganMoran) [inline DEVELOPMENT.md:84]: "MacOs" -> "MacOS"

File: Makefile [hunk 2]
```diff
@@ -27,14 +27,12 @@ clean:
 generate:
 	docker build ./scripts -f ./scripts/operator-sdk.docker -t 'operator-sdk:old'
 	go mod vendor
```
Comment (seanmalloy) [inline Makefile:29]: nit: we should consider removing the use of `go mod vendor` in the future.

File: deploy/crds/eunomia.kohls.io_gitopsconfigs_crd.yaml [hunk 3]
```diff
@@ -0,0 +1,156 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (seanmalloy) [inline deploy/crds/eunomia.kohls.io_gitopsconfigs_crd.yaml:1]: If this is the new CRD yaml file then the old CRD yaml file, eunomia_v1alpha1_gitopsconfig_crd.yaml, should be deleted.

File: pkg/apis/eunomia/v1alpha1/gitopsconfig_types.go [hunk 4]
```diff
@@ -7,9 +7,9 @@ import (
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (MeganMoran) [inline pkg/apis/eunomia/v1alpha1/gitopsconfig_types.go:7]: " THIS IS SCAFFOLDING FOR YOU TO OWN!" -> what does this mean?

File: pkg/controller/gitopsconfig/controller.go [hunk 5]
```diff
@@ -563,19 +570,18 @@ func ownedCronJobs(ctx context.Context, kube client.Client, owner *gitopsv1alpha
...
 	}
-	err = kube.List(ctx, &client.ListOptions{
-		LabelSelector: labels.NewSelector().Add(*filter),
-		Namespace:     owner.Namespace,
-	}, &jobs)
+	err := kube.List(ctx, jobs, listOpts...)
 	if err != nil {
 		return nil, fmt.Errorf("failed to list jobs for jobOwner==%q (ns: %s): %w", owner.Name, owner.Namespace, err)
 	}
+	for _, j := range jobs.Items {
```
Comment (MeganMoran) [inline pkg/controller/gitopsconfig/controller.go:582]: Change "j" to "job" for clarity

File: scripts/operator-sdk.docker [hunk 6]
```diff
@@ -13,5 +13,19 @@ VOLUME /gobase
 VOLUME /gocache
 VOLUME /gopath/src/github.com/KohlsTechnology/eunomia
 WORKDIR /gopath/src/github.com/KohlsTechnology/eunomia
-CMD echo "generating openapi..." && operator-sdk --verbose generate openapi && echo "generating k8s..." && operator-sdk --verbose generate k8s
+
+CMD \
+    go build -o /tmp/openapi-gen k8s.io/kube-openapi/cmd/openapi-gen && \
```
Comment (MeganMoran) [inline scripts/operator-sdk.docker:18]: Add echo "echo "generating openapi...""

Comment (MeganMoran) [inline scripts/operator-sdk.docker:18]: Hard code version of openapi-gen, if possible

Comment (Smiley73) [reply scripts/operator-sdk.docker:18]: It's actually using the version that was previously downloaded via `go mod vendor`. This version is already locked in via `go.mod`.

**Comment by Smiley73** [comment]
Lovely... We'll have to keep the old version of the CRD around for K8s < 1.16.

**Comment by seanmalloy** [comment]
> Lovely... We'll have to keep the old version of the CRD around for K8s < 1.16.

So how do we re-generate the CRD for older k8s versions? Are we saying it is not possible to do this if we upgrade to the latest sdk version?

What happens when we start working on the `v1alpha2` API? This API will require k8s 1.16?

**Comment by Smiley73** [comment]
> > Lovely... We'll have to keep the old version of the CRD around for K8s < 1.16.
> 
> So how do we re-generate the CRD for older k8s versions? Are we saying it is not possible to do this if we upgrade to the latest sdk version?
> 
> What happens when we start working on the `v1alpha2` API? This API will require k8s 1.16?

There's a good chance v1alpha2 will require K8s 1.16 and above. That being said, the changes between the 2 CRDs are minimal currently. There’s a good chance we just have to remove a couple of lines and it’ll work on older versions.

**Comment by seanmalloy** [comment]
@Smiley73 can you pin the `openapi-gen` version to not require `x-kubernetes-list-type`? This would allow the CRD to work with k8s versions less than 1.16.

**Comment by Smiley73** [comment]
On-hold until we're ready to support k8s > 1.16.
#345 has been implemented in the meantime

**Comment by seanmalloy** [comment]
Superseded by #374

**Pull request closed without merging**